# Modules
If you quit from the Python interpreter and enter it again, the definitions you have made (functions and variables) are lost. Therefore, if you want to write a somewhat longer program, you are better off using a text editor to prepare the input for the interpreter and running it with that file as input instead. This is known as creating a script. As your program gets longer, you may want to split it into several files for easier maintenance. You may also want to use a handy function that you’ve written in several programs without copying its definition into each program.

To support this, Python has a way to put definitions in a file and use them in a script or in an interactive instance of the interpreter. Such a file is called a module; definitions from a module can be imported into other modules or into the main module (the collection of variables that you have access to in a script executed at the top level and in calculator mode).

A module is a file containing Python definitions and statements. The file name is the module name with the suffix .py appended. Within a module, the module’s name (as a string) is available as the value of the global variable __name__. For instance, use your favorite text editor to create a file called fibo.py in the current directory with the following contents:


In [5]:
import fibon

In [12]:
fibon.fib2(25)

[0, 1, 1, 2, 3, 5, 8, 13, 21]

In [13]:
fib11 = fibon.fib
fib11(500)

0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 


# More on Modules

# A module can contain executable statements as well as function definitions. 

# These statements are intended to initialize the module. 

# They are executed only the first time the module name is encountered in an import statement. 

# (They are also run if the file is executed as a script.)

# Each module has its own private namespace, which is used as the global namespace by all functions defined in the module. 

# Thus, the author of a module can use global variables in the module without worrying about accidental clashes with a user’s global variables. On the other hand, if you know what you are doing you can touch a module’s global variables with the same notation used to refer to its functions, modname.itemname.

# Modules can import other modules. 

# It is customary but not required to place all import statements at the beginning of a module (or script, for that matter). 

# The imported module names, if placed at the top level of a module (outside any functions or classes), 
# are added to the module’s global namespace.

# There is a variant of the import statement that imports names from a module directly into the importing module’s namespace.

# For example:


# from fibo import fib, fib2

# fib(500)

# 0 1 1 2 3 5 8 13 21 34 55 89 144 233 377

# This does not introduce the module name from which the imports are taken in the local namespace 

# (so in the example, fibo is not defined).

 
# There is even a variant to import all names that a module defines:


# from fibo import *

# fib(500)

# 0 1 1 2 3 5 8 13 21 34 55 89 144 233 377

# This imports all names except those beginning with an underscore (_). 

# In most cases Python programmers do not use this facility since it introduces an unknown set of names into the interpreter, 

# possibly hiding some things you have already defined.

# Note that in general the practice of importing * from a module or package is frowned upon, 

# since it often causes poorly readable code. However, it is okay to use it to save typing in interactive sessions.

# If the module name is followed by as, then the name following as is bound directly to the imported module.


# import fibo as fib

# fib.fib(500)

# 0 1 1 2 3 5 8 13 21 34 55 89 144 233 377

# This is effectively importing the module in the same way that import fibo will do, with the only difference of it 

# being available as fib.

# It can also be used when utilising from with similar effects:

# from fibo import fib as fibonacci

# fibonacci(500)

# 0 1 1 2 3 5 8 13 21 34 55 89 144 233 377

# Note For efficiency reasons, each module is only imported once per interpreter session. 

# Therefore, if you change your modules, you must restart the interpreter – or, 

# if it’s just one module you want to test interactively, 

# use importlib.reload(), e.g. import importlib; importlib.reload(modulename).

In [ ]:
from fibon import fib, fib2
fib(500)

In [15]:
from fibo import fib, fib2
fib(500)

0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 


In [16]:
from fibo import *
fib(500)

0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 


In [17]:
import fibo as fib
fib.fib(500)

0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 


In [18]:
from fibo import fib as fibonacci
fibonacci(500)

0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 


In [19]:
import calc as cl

In [20]:
cl.add(5,6)

11

In [21]:
cl.subtract(7,2)

5

In [22]:
# Python Import From Module
# Python’s from statement lets you import specific attributes from a module without importing the module as a whole.

# Import Specific Attributes from a Python module
# Here, we are importing specific sqrt and factorial attributes from the math module.


# importing sqrt() and factorial from the
# module math
from math import sqrt, factorial
 
# if we simply do "import math", then
# math.sqrt(16) and math.factorial()
# are required.
print(sqrt(16))
print(factorial(6))

4.0
720


In [23]:
# What does import * do in Python
# The use of * has its advantages and disadvantages. 
# If you know exactly what you will be needing from the module, it is not recommended to use *, else do so.


# importing sqrt() and factorial from the
# module math
from math import *
 
# if we simply do "import math", then
# math.sqrt(16) and math.factorial()
# are required.
print(sqrt(16))
print(factorial(6))

4.0
720


In [24]:
# Locating Python Modules
#Whenever a module is imported in Python the interpreter looks for several locations. 
#First, it will check for the built-in module, if not found then it looks for a list of directories defined in the sys.path. 
#Python interpreter searches for the module in the following manner –

#First, it searches for the module in the current directory.
#If the module isn’t found in the current directory, Python then searches each directory in the shell variable PYTHONPATH. 
#The PYTHONPATH is an environment variable, consisting of a list of directories.
#If that also fails python checks the installation-dependent list of directories configured at the time Python is installed.
#Directories List for Modules
#Here, sys.path is a built-in variable within the sys module.
#It contains a list of directories that the interpreter will search for the required module.


# importing sys module
import sys
 
# importing sys.path
print(sys.path)

['C:\\Users\\soham', 'C:\\Users\\soham\\anaconda3\\python38.zip', 'C:\\Users\\soham\\anaconda3\\DLLs', 'C:\\Users\\soham\\anaconda3\\lib', 'C:\\Users\\soham\\anaconda3', '', 'C:\\Users\\soham\\anaconda3\\lib\\site-packages', 'C:\\Users\\soham\\anaconda3\\lib\\site-packages\\locket-0.2.1-py3.8.egg', 'C:\\Users\\soham\\anaconda3\\lib\\site-packages\\win32', 'C:\\Users\\soham\\anaconda3\\lib\\site-packages\\win32\\lib', 'C:\\Users\\soham\\anaconda3\\lib\\site-packages\\Pythonwin', 'C:\\Users\\soham\\anaconda3\\lib\\site-packages\\IPython\\extensions', 'C:\\Users\\soham\\.ipython']


In [25]:
#Python Built-in modules
#There are several built-in modules in Python, which you can import whenever you like.
# importing built-in module math
import math

# using square root(sqrt) function contained 
# in math module
print(math.sqrt(25)) 

# using pi function contained in math module
print(math.pi) 

# 2 radians = 114.59 degrees
print(math.degrees(2)) 

# 60 degrees = 1.04 radians
print(math.radians(60)) 

# Sine of 2 radians
print(math.sin(2)) 

# Cosine of 0.5 radians
print(math.cos(0.5)) 

# Tangent of 0.23 radians
print(math.tan(0.23)) 

# 1 * 2 * 3 * 4 = 24
print(math.factorial(4)) 

# importing built in module random
import random

# printing random integer between 0 and 5
print(random.randint(0, 5)) 

# print random floating point number between 0 and 1
print(random.random()) 

# random number between 0 and 100
print(random.random() * 100) 

List = [1, 4, True, 800, "python", 27, "hello"]

# using choice function in random module for choosing 
# a random element from a set such as a list
print(random.choice(List)) 


# importing built in module datetime
import datetime
from datetime import date
import time

# Returns the number of seconds since the
# Unix Epoch, January 1st 1970
print(time.time()) 

# Converts a number of seconds to a date object
print(date.fromtimestamp(454554)) 





5.0
3.141592653589793
114.59155902616465
1.0471975511965976
0.9092974268256817
0.8775825618903728
0.23414336235146527
24
2
0.38686470995725253
68.41951080768685
800
1698640459.5975049
1970-01-06
